src/skadi/sample.cc

```diff
@@ -442,18 +439,89 @@ template <class coords_t> std::vector<double> sample::get_all(const coords_t& co
       curIndex = index;
     }
 
-    double value = NO_DATA_VALUE;
+    double value{get_no_data_value()};
     if (tile) {
       double u = (coord.first - lon) * (HGT_DIM - 1);
       double v = (1.0 - (coord.second - lat)) * (HGT_DIM - 1);
       value = tile.get(u, v);
     }
 
+    if (value == get_no_data_value())
+      value = get_from_remote(coord);
+
     values.emplace_back(value);
   }
+
   return values;
 }
 
+template <class coord_t> double sample::get_from_cache(const coord_t& coord) {
+  auto tile = cache_->source(get_tile_index(coord));
+  if (!tile)
+    return get_no_data_value();
+
+  double u = (coord.first - std::floor(coord.first)) * (HGT_DIM - 1);
+  double v = (1.0 - (coord.second - std::floor(coord.second))) * (HGT_DIM - 1);
+  return tile.get(u, v);
+}
+
+bool sample::store(const std::string& elev, const std::vector<char>& raw_data) {
+  auto fpath = cache_->data_source + elev;
+  if (filesystem::exists(fpath))
+    return true;
+
+  auto data = cache_item_t::parse_hgt_name(elev);
+  if (!data || data->second == format_t::UNKNOWN)
+    return false;
+
+  if (data->first >= cache_->cache.size())
+    return false;
+
+  if (!filesystem::save(fpath, raw_data))
+    return false;
+  auto res = cache_->cache[data->first].init(fpath, data->second);
+  return res;
+}
+
+template <class coord_t> double sample::get_from_remote(const coord_t& coord) {
+  if (url_.empty() || !remote_loader_)
+    return get_no_data_value();
+
+  auto index = get_tile_index(coord);
+  auto elev = get_hgt_file_name(index);
+  auto uri = baldr::make_single_point_url(url_, elev, remote_path_);
+
+  {
+    std::lock_guard<std::mutex> _(st_lck_);
+    if (st_.count(uri))
+      return get_from_cache(coord);
+
+    st_.insert(uri);
+  }
```

Comment:
It is not clear how all these `get_from_remote` and `get_from_cache` works together
I'll come back to this thread soon, I need to think about it...

Reply:
I think the current implementation doesn't handle a case correctly when two threads ask for the same tile
For example
- 1 thread called `get_from_remote`. There's no such tile on the filesystem so it started fetching it.
- 2 thread called `get_from_remote`. The check above(`st.count(uri)`) is passed as the first thread already added such key. However, the tile has been fetched completely yet

So we need some kind of blocking for the same tile. Maybe something similar to what we have at the moment(I mean `pending_tiles` https://github.com/valhalla/valhalla/blob/dd8699dc8436ec9da69c5dc6925f51b65471febb/src/skadi/sample.cc#L258)

Reply:
i think i found different solution at the end but let's discuss this one either.

Reply:
I think there are a few main points I'm trying to say
- If multiple threads want to fetch the same tile - only one thread initiates fetching and other threads should wait(block) for the result
- If I'm not mistaken the current (master) implementation doesn't assume  that tiles suddenly land on the filesystem. So we need to adopt current implementation for on-demand use case and make it thread safe
- Make sure everything else is thread safe and wrapped with synchronization(mutex-es) if needed